proposals/NNNN-uniformity-qualifiers.md

```diff
@@ -0,0 +1,131 @@
+the importance of this fundamental property it is not represented in any
+explicit way in the HLSL language.
+
+This proposal, seeks to address that by introducing core concepts around
+uniformity to HLSL's type system and programming model.
+
+## Proposed solution
+
+### Uniformity as a Type Qualifier
+
+HLSL should introduce a new set of type qualifiers to represent the different
+scopes of uniformity:
+* `group_uniform`
+* `simd_uniform`
+* `quad_uniform`
+* non-uniform (default state with no associated keyword)
+
+`group_uniform` is the highest scoping of uniformity, and implies all other
+scopes. `simd_uniform` implies `quad_uniform`.
+
+A new "UniformityReduction" cast will reduce the uniform scope allowing
+conversion of one uniformity scope to another uniformity scope as long as the
+source scope has a greater uniformity scope.
+
+Any GLValue with a uniformity scope can be implicitly converted to a GLValue
+with reduced uniform scope or no uniformity scope.
+
+Any PRValue with a uniformity scope can be implicitly converted to a PRValue
+with reduced uniform scope or no uniformity scope.
+
+No implicit or explicit cast can increase uniformity scope.
+
+HLSL library functionality that produces uniform results will be updated to
+produce appropriately qualified uniform types. These functions can produce
+uniform values from non-uniform inputs. For example:
+
+```hlsl
+simd_uniform bool WaveActiveAllTrue(bool);
+quad_uniform bool QuadAny(bool);
+```
+
+Compile-time constants and Groupshared variable declarations imply
+`group_uniform` uniformity.
+
+Builtin operators will produce uniform result values based on the uniformity of
+the intersection of the uniformity of the arguments.
+
+Vector and matrix component access expressions and structure member expressions
+will produce result values with the same uniformity of the base object.
+
+```hlsl
+groupshared int SomeData[10];
+simd_uniform int WaveReadLaneFirst(int);
+
+void fn(int Val) {
+  simd_uniform Idx = WaveReadLaneFirst(Val); // produces a simd_uniform value.
+  // group_uniform indexed by simd_uniform produces simd_uniform value.
+  simd_uniform GSVal = SomeData[Idx];
+  // group_uniform with group_uniform index produces group_uniform value.
+  group_uniform GSVal2 = SomeData[SomeData[0]];
+  // Binary operator of group_uniform and simd_uniform values produces a
+  // simd_uniform value.
+  if (GSVal > GSVal2) { // This control-flow can be defined as simd_uniform
+
+  }
+}
+```
+
+Uniformity qualifiers may be applied on shader inputs. When applied to an input
+the compiler will diagnose known cases where the qualifier mismatches, and it
+will trust the user in other cases. A runtime validation may be added to catch
+incorrect source annotations.
+
+### Uniformity Requirements for Functions
+
+HLSL should introduce a new set of attributes for defining the control flow
+uniformity requirements of functions. These new attributes should take the form:

```

Comment:
Wondering if we should try and frame this as what this proposal is proposing, rather than what we think someone needs to propose?

```suggestion
This proposal introduces a new set of attributes for defining the control flow
uniformity requirements of functions. These new attributes take the form:
```